ToolsForCategoricalTowers/gap/Tools.gi

```diff
@@ -107,6 +107,224 @@ InstallMethod( GlobalSectionFunctor,
+        
+    fi;
+    
+end );
+
+##
+InstallMethod( CellAsEvaluatableString,
+        [ IsCapCategoryObject, IsList ],
+
+  function( obj, list_of_evaluatable_strings )
+    local list_of_values, C, pos, string;
+    
+    list_of_values := List( list_of_evaluatable_strings, EvalString );
+    
+    C := CapCategory( obj );
+    
+    pos := PositionsProperty( list_of_values, val ->
+                   IsCapCategoryObject( val ) and IsIdenticalObj( CapCategory( val ), C ) and IsEqualForObjects( val, obj ) );
+    
+    if Length( pos ) = 1 then
+        
+        return list_of_evaluatable_strings[pos[1]];
+        
+    elif Length( pos ) > 1 then
+        
+        Error( "the position of the object `obj` in `list_of_values` is not unique for obj = ", obj, "\n" );
+        
+    fi;
+    
+    pos := PositionsProperty( list_of_values, val -> IsIdenticalObj( val, C ) );
+    
+    if Length( pos ) = 0 then
+        
+        Error( "could not find CapCategory( obj ) in `list_of_values` for obj = ", obj, "\n" );
+        
+    elif Length( pos ) > 1 then
+        
+        Error( "the position of CapCategory( obj ) in `list_of_values` is not unique for obj = ", obj, "\n" );
+        
+    fi;
+    
+    string := Concatenation(
+                      "ObjectConstructor( ", list_of_evaluatable_strings[pos[1]], ", ",
+                      DatumOfCellAsEvaluatableString( obj, list_of_evaluatable_strings ), " )" );
+    
+    #Assert( 0, IsEqualForObjects( obj, EvalString( string ) ) );
+    
+    return string;
+    
+end );
+
+##
+InstallMethod( CellAsEvaluatableString,
+        [ IsCapCategoryMorphism, IsList ],
+
+  function( mor, list_of_evaluatable_strings )
+    local list_of_values, C, pos, cat, string;
+    
+    list_of_values := List( list_of_evaluatable_strings, EvalString );
+    
+    C := CapCategory( mor );
+    
+    pos := PositionsProperty( list_of_values, val ->
+                   IsCapCategoryMorphism( val ) and IsIdenticalObj( CapCategory( val ), C ) and IsEqualForMorphismsOnMor( C, val, mor ) );
+    
+    if Length( pos ) = 1 then
+        
+        return list_of_evaluatable_strings[pos[1]];
+        
+    elif Length( pos ) > 1 then
+        
+        Error( "the position of the morphism `mor` in `list_of_values` is not unique for mor = ", mor, "\n" );
+        
+    fi;
+    
+    pos := PositionsProperty( list_of_values, val -> IsIdenticalObj( val, C ) );
+    
+    if Length( pos ) = 0 then
+        
+        Error( "could not find CapCategory( mor ) in `list_of_values` for mor = ", mor, "\n" );
+        
+    elif Length( pos ) > 1 then
+        
+        Error( "the position of CapCategory( mor ) in `list_of_values` is not unique for mor = ", mor, "\n" );
+        
+    fi;
+    
+    cat := list_of_evaluatable_strings[pos[1]];
+    
+    if CanCompute( C, "IsEqualToIdentityMorphism" ) and IsEqualToIdentityMorphism( mor ) then
```

Comment:
Do you have an application where this special case (or the zero case below) is needed?

I don't expect this case to be needed. Previously I thought it would be needed, due to our experiments with FreeCategory, where we could not properly handle morphism data. However, now I think this simply is a deficiency with FreeCategory. With PathCategory, the identity morphism given by the empty list is easily expressible in primitive data structures. Moreover, the ur-algorithm can actually evaluate this primitive data structure without further help, while evaluating a functor on the literal expression `IdentityMorphism( ... )` during compilation requires additional logic templates.

Reply:
Excellent point :)

> Do you have an application where this special case (or the zero case below) is needed?

No :)

`IdentityMorphism( ... )` just looks nicer than `PreComposeList( ..., [ ], ... )` :)
 
> I don't expect this case to be needed.

Correct :)

> Previously I thought it would be needed, due to our experiments with FreeCategory, where we could not properly handle morphism data.

Exactly. This was the reason for having a special treatment for `IsEqualToIdentityMorphism`

> However, now I think this simply is a deficiency with FreeCategory. With PathCategory, the identity morphism given by the empty list is easily expressible in primitive data structures.

Yes :)

> Moreover, the ur-algorithm can actually evaluate this primitive data structure without further help, while evaluating a functor on the literal expression `IdentityMorphism( ... )` during compilation requires additional logic templates.

Absolutely correct :)

In fact, for this PR I deliberately changed `PreComposeList( ..., [ ], ... )` to `IdentityMorphism( ... )` for the output to be more readable. In the PR with the various extension functors (call the various ur-algorithms), I will make `PreComposeList( ..., [ ], ... )` the default behavior and `IdentityMorphism( ... )` (and `ZeroMorphism( ... )`) reachable through an option.

Reply:
Update:

In the various `ExtendFunctorTo*` for `IsPathCategory`, `IsFpCategory`, and `IsCategoryFromDataTables` I indeed use `PreComposeList( source, MorphismDatum( ... ), target )`. All `ExtendFunctorTo*`, including the already implemented ones, use `MorphismDatum( ... )` before applying the functor to be extended. This means that `IdentityMorphism( ... )` ultimately gets resolved into code containing `PreComposeList( ..., [ ], ... )`. And the wonderful compiler finally uses the standard derivation of `PreComposeList` to resolve `PreComposeList( ..., [ ], ... )` into `IdentityMorphism( ... )`  in the target the category :)

Reply:
> This means that `IdentityMorphism( ... )` ultimately gets resolved into code containing `PreComposeList( ..., [ ], ... )`.

Ah, I did not think of this. So we do not need a manual logic template, perfect.